src/main/java/uk/ac/ebi/eva/pipeline/io/readers/VcfHeaderReader.java

```diff
@@ -100,12 +104,22 @@ public VariantSourceEntity read() throws Exception {
         }
     }
 
-    private VariantSourceEntity doRead() {
+    private VariantSourceEntity doRead() throws DuplicateSamplesFoundException {
         if (variantReader == null) {
             throw new IllegalStateException("The method VcfHeaderReader.open() should be called before reading");
         }
         variantReader.pre();
         source.addMetadata(VARIANT_FILE_HEADER_KEY, variantReader.getHeader());
+
+        List<String> sampleNames = new ArrayList<String> ( variantReader.getSampleNames());
+        HashSet<String> uniqueSampleNames = new HashSet<String>(sampleNames);
+
+        if (sampleNames.size() != uniqueSampleNames.size()) {
+            for (String str : uniqueSampleNames) {
+              sampleNames.remove(str);
```

Comment:
While this is functional, the performance complexity of remove is O(n). Since this remove operation is inside a loop, complexity here is O(n*n). This can escalate quickly even for studies with moderately large number of samples (say 1000 genomes). @cyenyxe Any thoughts?

Reply:
Fair point! Since a HashSet is usually backed by a HashMap I could indeed just check for duplication while inserting into the HashSet rather than by removing from the slower ArrayList.
HashSet's add(obj) returns a status boolean showing whether the object was a duplicate or not. So I can just iteratively add the sample names rather than use a copy constructor and then check for any true flags returned at the end. Does that sound reasonable?